source/common/http/filter_manager.cc

```diff
@@ -903,16 +903,28 @@ FilterManager::commonEncodePrefix(ActiveStreamEncoderFilter* filter, bool end_st
                                   FilterIterationStartState filter_iteration_start_state) {
   // Only do base state setting on the initial call. Subsequent calls for filtering do not touch
   // the base state.
-  ENVOY_STREAM_LOG(trace, "commonEncodePrefix end_stream: {}, isHalfCloseEnabled: {}", *this,
-                   end_stream, filter_manager_callbacks_.isHalfCloseEnabled());
+  ENVOY_STREAM_LOG(trace,
+                   "commonEncodePrefix end_stream: {}, isHalfCloseEnabled: {}, force_close: {}",
+                   *this, end_stream, filter_manager_callbacks_.isHalfCloseEnabled(),
+                   static_cast<bool>(state_.force_close_stream_));
   if (filter == nullptr) {
     // half close is enabled in case tcp proxying is done with http1 encoder. In this case, we
     // should not set the local_complete_ flag to true when end_stream is true.
     // setting local_complete_ to true will cause any data sent in the upstream direction to be
     // dropped.
-    if (end_stream && !filter_manager_callbacks_.isHalfCloseEnabled()) {
-      ASSERT(!state_.local_complete_);
-      state_.local_complete_ = true;
+    if (allow_upstream_half_close_) {
+      if (end_stream) {
+        state_.encoder_end_stream_ = true;
+        if (!filter_manager_callbacks_.isHalfCloseEnabled() || state_.force_close_stream_) {
+          ASSERT(!state_.local_complete_);
+          state_.local_complete_ = true;
+        }
+      }
+    } else {
+      if (end_stream && !filter_manager_callbacks_.isHalfCloseEnabled()) {
```

Comment:
I feel like we should be able to share code between these two

Reply:
My goal here was to make it clear that the old behavior is fully preserved when `allow_upstream_half_close_ == false`. Refactor can save a couple of lines of code but will make it harder to see whether behavior has changed or not.

Reply:
I'm concerned that we have allow_upstream_half_close and isHalfCloseEnabled.
if filter_manager_callbacks_.isHalfCloseEnabled isn't working as intended can we just fix?

Reply:
I have changed FM to always use filter_manager_callbacks_.isHalfCloseEnabled

I will refactor FM to just always support independent half closes when runtime flag is removed.